onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc

```diff
@@ -0,0 +1,139 @@
+      continue;
+    }
+
+    Node& cast_node = const_cast<Node&>(left_edges[0]->GetNode());
+    Node& matmulinteger_node = const_cast<Node&>(left_edges[1]->GetNode());
+    Node& dql_node_left = const_cast<Node&>(left_edges[2]->GetNode());
+
+    Node& mul_node_right = const_cast<Node&>(right_edges[0]->GetNode());
+    Node& dql_node_right = const_cast<Node&>(right_edges[1]->GetNode());
+
+    // Check if left DynamicQuantizeLinear is same as right DynamicQuantizeLinear
+    if (dql_node_left.Index() != dql_node_right.Index()) {
+      continue;
+    }
+
+    // Check Nodes' Edges count and Nodes' outputs are not in Graph output
+    if (cast_node.GetOutputEdgesCount() != 1 ||
+        matmulinteger_node.GetOutputEdgesCount() != 1 ||
+        mul_node_right.GetOutputEdgesCount() != 1 ||
+        dql_node_left.GetOutputEdgesCount() != 3 ||
+        !graph.GetNodeOutputsInGraphOutputs(cast_node).empty() ||
+        !graph.GetNodeOutputsInGraphOutputs(matmulinteger_node).empty() ||
+        !graph.GetNodeOutputsInGraphOutputs(mul_node_right).empty() ||
+        !graph.GetNodeOutputsInGraphOutputs(dql_node_left).empty()) {
+      continue;
+    }
+
+    const NodeArg& matmulinteger_B = *(matmulinteger_node.InputDefs()[1]);
+    if (!graph_utils::IsInitializer(graph, matmulinteger_B.Name(), true)) {
+      continue;
+    }
+
+    const NodeArg& mul_right_B = *(mul_node_right.InputDefs()[1]);
+    if (!graph_utils::IsInitializer(graph, mul_right_B.Name(), true)) {
+      continue;
+    }
+
+    std::vector<NodeArg*> input_defs{dql_node_left.MutableInputDefs()[0],
+                                     matmulinteger_node.MutableInputDefs()[1],
+                                     mul_node_right.MutableInputDefs()[1]};
+
+    if (matmulinteger_node.InputDefs().size() == 4) {
+      const NodeArg& matmulinteger_B_zp = *(matmulinteger_node.InputDefs()[3]);
+      if (!graph_utils::IsInitializer(graph, matmulinteger_B_zp.Name(), true)) {
+        continue;
+      }
+      input_defs.push_back(matmulinteger_node.MutableInputDefs()[3]);
+    }
+
+    Node& fused_node = graph.AddNode(graph.GenerateNodeName("DynamicQuantizeMatMul"),
+                                     "DynamicQuantizeMatMul",
+                                     "fused DynamicQuantizeMatMul",
+                                     input_defs,
+                                     mul_node.MutableOutputDefs(),
+                                     nullptr,
+                                     kMSDomain);
+
+    // Assign provider to this new node. Provider should be same as the provider for old node.
+    fused_node.SetExecutionProviderType(mul_node.GetExecutionProviderType());
+
+    nodes_to_remove.push_back(dql_node_left);
+    nodes_to_remove.push_back(matmulinteger_node);
+    nodes_to_remove.push_back(cast_node);
+    nodes_to_remove.push_back(mul_node_right);
+    nodes_to_remove.push_back(mul_node);
+  }
```

Comment:
graph_utils::FinalizeNodeFusion should be able to do all this and you don't need to keep a list of nodes_to_remove. Don't provide output defs in the AddNode call, and make sure mul_node is last in the list of nodes used in the call to FinalizeNodeFusion (as it will move the output defs from mul_node to the new node).

Reply:
I would like to keep this one. It is a little bit trick with FinalizeNodeFusion because it just override the first input of the fused op. 

---
In reply to: [439086726](https://github.com/microsoft/onnxruntime/pull/4195#discussion_r439086726) [](ancestors = 439086726)